linalg/tadvector.hpp

```diff
@@ -0,0 +1,703 @@
+
+   /// Changes the ownership of the data; after the call the Vector is empty
+   inline void StealData(dtype **p)
+   {
+      *p = data;
+      delete [] data;
+      size = 0;
+      capacity = 0;
+   }
+
+   /// Changes the ownership of the data; after the call the Vector is empty
+   inline dtype *StealData()
+   {
+      dtype *p;
+      StealData(&p);
+      return p;
+   }
+
+   /// Access Vector entries. Index i = 0 .. size-1.
+   dtype &Elem(int i)
+   {
+      return operator()(i);
+   }
+   /// Read only access to Vector entries. Index i = 0 .. size-1.
+   const dtype &Elem(int i) const
+   {
+      return operator()(i);
+   }
+
+   /// Access Vector entries using () for 0-based indexing.
+   /** @note If MFEM_DEBUG is enabled, bounds checking is performed. */
+   inline dtype &operator()(int i)
+   {
+      MFEM_ASSERT(data && i >= 0 && i < size,
+                  "index [" << i << "] is out of range [0," << size << ")");
+
+      return data[i];
+   }
+
+   /// Read only access to Vector entries using () for 0-based indexing.
+   /** @note If MFEM_DEBUG is enabled, bounds checking is performed. */
+   inline const dtype &operator()(int i) const
+   {
+      MFEM_ASSERT(data && i >= 0 && i < size,
+                  "index [" << i << "] is out of range [0," << size << ")");
+
+      return data[i];
+   }
+
+   /// Dot product with a `dtype *` array.
+   dtype operator*(const dtype *v) const
+   {
+      dtype dot = 0.0;
+#ifdef MFEM_USE_LEGACY_OPENMP
+      #pragma omp parallel for reduction(+:dot)
+#endif
+      for (int i = 0; i < size; i++)
+      {
+         dot += data[i] * v[i];
+      }
+      return dot;
+   }
+
+
+   /// Return the inner-product.
+   dtype operator*(const TADVector<dtype> &v) const
+   {
+      MFEM_ASSERT(size == v.Size(), "incompatible Vectors!");
+      dtype dot = 0.0;
+      for (int i = 0; i < size; i++)
+      {
+         dot += data[i] * v[i];
+      }
+      return dot;
+   }
+
+   dtype operator*(const Vector &v) const
+   {
+      MFEM_ASSERT(size == v.Size(), "incompatible Vectors!");
+      dtype dot = 0.0;
+      for (int i = 0; i < size; i++)
+      {
+         dot += data[i] * v[i];
+      }
+      return dot;
+   }
+
+   /// Copy Size() entries from @a v.
+   TADVector<dtype> &operator=(const dtype *v)
+   {
+      for (int i=0; i<size; i++)
+      {
+         data[i]=v[i];
+      }
+      return *this;
+   }
+
+   /// Copy assignment.
+   /** @note Defining this method overwrites the implicitly defined copy
+       assignemnt operator. */
+   TADVector<dtype> &operator=(const TADVector<dtype> &v)
+   {
+      SetSize(v.Size());
+      for (int i=0; i<size; i++)
+      {
+         data[i]=v[i];
+      }
+      return *this;
+   }
+
+   TADVector<dtype> &operator=(const Vector &v)
+   {
+      SetSize(v.Size());
+      for (int i=0; i<size; i++)
+      {
+         data[i]=v[i];
+      }
+      return *this;
+   }
+
+   /// Redefine '=' for vector = constant.
+   template<typename ivtype>
```

Comment:
they could be double and float for example - templates will automatically define the method for different types. I can make them to be only for the base type - let me know if you think that it will be better. 

I do not think that MFEM has style guide for these cases.